File: 0028_Implement_strStr/Implement_strStr.rb
# @param {String} haystack
# @param {String} needle
# @return {Integer}
def str_str(haystack, needle)
  return 0 if needle.empty?
  return -1 if haystack && needle && haystack[needle].nil?
  (0...haystack.length).each do |i|
    (0...needle.length).each do |j|
      break if haystack[i + j] != needle[j]
      return i if j == needle.length - 1
    end
  end
end

File: 0028_Implement_strStr/Implement_strStr.py
# -*- coding: utf-8 -*-

"""
Implement strStr().

Return the index of the first occurrence of needle in haystack, or -1 if needle is not part of haystack.

Example 1:

Input: haystack = "hello", needle = "ll"
Output: 2

Example 2:

Input: haystack = "aaaaa", needle = "bba"
Output: -1

Clarification:

What should we return when needle is an empty string? This is a great question to ask during an interview.

For the purpose of this problem, we will return 0 when needle is an empty string. This is consistent to C's strstr() and Java's indexOf().
"""


class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        if not needle:
            return 0
        if not (haystack and needle and (needle in haystack)):
            return -1
        for i in range(len(haystack)):
            for j in range(len(needle)):
                if haystack[i + j] != needle[j]:
                    break
                if j == len(needle) - 1:
                    return i
